readme.md
# tk-mail

__Project:__ tk-mail  
__Web:__ <https://github.com/tropotek/tk-mail/>  
__Authors:__ Michael Mifsud <http://www.tropotek.com/>  
  
A mail lib for the Tk lib. Uses PHPMailer for the email driver.

## Contents

- [Installation](#installation)
- [Introduction](#introduction)


## Installation

Available on Packagist ([ttek/tk-mail](http://packagist.org/packages/ttek/tk-mail))
and installable via [Composer](http://getcomposer.org/).

```bash
composer require ttek/tk-mail
```

Or add the following to your composer.json file:

```json
{
  "require": {
    "ttek/tk-mail": "~8.0.0"
  }
}
```


## Introduction

Basic Example:
```php
$message = new \Tk\Mail\Message();
$message->addTo('info@tropotek.com');
$message->setFrom('godar@tropotek.com.au');
$message->setSubject('This is a test email');
$message->setBody(\App\Config::createMailTemplate('This is some message text'));

$message->send();
```

Available Config Params:

```php
return function (\Tk\Config $config)
{
    /*
     * Options (php)mail, smtp, sendmail, qmail
     */
    $config->set('mail.driver', 'mail');

    /*
     * SMTP settings
     */

    /**
     * SMTP hosts.
     * Either a single hostname or multiple semicolon-delimited hostnames.
     * You can also specify a different port
     * for each host by using this format: [hostname:port]
     * (e.g. "smtp1.example.com:25;smtp2.example.com").
     * You can also specify encryption type, for example:
     * (e.g. "tls://smtp1.example.com:587;ssl://smtp2.example.com:465").
     * Hosts will be tried in order.
     */
    $config->set('mail.smtp.host', 'localhost');
    $config->set('mail.smtp.port', 25);
    $config->set('mail.smtp.username', '');
    $config->set('mail.smtp.password', '');
    // What kind of encryption to use on the SMTP connection. Options: '', 'ssl' or 'tls'
    $config->set('mail.smtp.secure', '');
    // Whether to use SMTP authentication. Uses the Username and Password properties.
    $config->set('mail.smtp.enableAuth', true);
    // Whether to keep SMTP connection open after each message.
    //   If this is set to true then to close the connection requires an explicit call to smtpClose().
    $config->set('mail.smtp.username', false);

    // Checks if the send command was called from this site
    $config->set('mail.checkReferer', true);
    // Add valid domain names as valid referrers if needed
    $config->set('mail.validReferrers', '');

    /*
     * Other misc options
     */

    // Set this if you want all email to go to this address in debug mode
    //$config->set('mail.debug.email', 'user@example.com');

    // If set X-Application will be set to this
    //$config->set('mail.name', 'tk-mail');
    //$config->set('mail.version', ''1.0.0);

    // Change this to suite your message body encoding
    $config->set('mail.encoding', 'UTF-8');


    // DKIM SETUP

    //$config->set('mail.dkim.domain', 'example.com');
    //$config->set('mail.dkim.private', '/path/to/my/private.key');
    //$config->set('mail.dkim.private_string', '');         // Use the private key string instead of a file


};
```

### DKIM setup

See: https://stackoverflow.com/questions/24463425/send-mail-in-phpmailer-using-dkim-keys
```
If you take a look in the PHPMailer unit tests, there is an example of how to set up DKIM.

Here are the basics beyond what you already need to do to send a message 
(obviously change the domain, key path and selector to match your config, and add a passphrase if you use one); 
this also assumes that you are intending to sign using the same identifier as your From address:

$mail->DKIM_domain = 'example.com';
$mail->DKIM_private = '/path/to/my/private.key';
$mail->DKIM_selector = 'phpmailer';
$mail->DKIM_passphrase = '';
$mail->DKIM_identity = $mail->From;

When you send() the message (and not before), it will use these settings to generate a DKIM signature.
```




changelog.md
#CHANGELOG#

Ver 3.2.22 [2023-12-19]:
-------------------------------
  - added new student DB table and fields
  - Tagging and releasing branch `3.2` with version `3.2.20`.
  - Tagging and releasing branch `3.2` with version `3.2.18`.
  - Only send disposal reminders to necropsy cases
  - Tagging and releasing branch `3.2` with version `3.2.16`.
  - Added option to ue text client name
  - Tagging and releasing branch `3.2` with version `3.2.14`.
  - Tagging and releasing branch `3.2` with version `3.2.12`.
  - Tagging and releasing branch `3.2` with version `3.2.10`.
  - Tagging and releasing branch `3.2` with version `3.2.8`.
  - Added DKIM signing to institution settings
  - Tagging and releasing branch `3.2` with version `3.2.6`.
  - Use replyTo instead of from for emails
  - Tagging and releasing branch `3.2` with version `3.2.4`.
  - Tagging and releasing branch `3.2` with version `3.2.2`.
  - Tagging and releasing branch `3.2` with version `3.2.0`.
  - Tagging and releasing branch `3.0` with version `3.0.26`.
  - Tagging and releasing branch `3.0` with version `3.0.24`.
  - Tagging and releasing branch `3.0` with version `3.0.22`.
  - Tagging and releasing branch `3.0` with version `3.0.20`.
  - Tagging and releasing branch `3.0` with version `3.0.18`.
  - Tagging and releasing branch `3.0` with version `3.0.14`.


Ver 3.2.20 [2023-11-18]:
-------------------------------
  - Tagging and releasing branch `3.2` with version `3.2.18`.
  - Only send disposal reminders to necropsy cases
  - Tagging and releasing branch `3.2` with version `3.2.16`.
  - Added option to ue text client name
  - Tagging and releasing branch `3.2` with version `3.2.14`.
  - Tagging and releasing branch `3.2` with version `3.2.12`.
  - Tagging and releasing branch `3.2` with version `3.2.10`.
  - Tagging and releasing branch `3.2` with version `3.2.8`.
  - Added DKIM signing to institution settings
  - Tagging and releasing branch `3.2` with version `3.2.6`.
  - Use replyTo instead of from for emails
  - Tagging and releasing branch `3.2` with version `3.2.4`.
  - Tagging and releasing branch `3.2` with version `3.2.2`.
  - Tagging and releasing branch `3.2` with version `3.2.0`.
  - Tagging and releasing branch `3.0` with version `3.0.26`.
  - Tagging and releasing branch `3.0` with version `3.0.24`.
  - Tagging and releasing branch `3.0` with version `3.0.22`.
  - Tagging and releasing branch `3.0` with version `3.0.20`.
  - Tagging and releasing branch `3.0` with version `3.0.18`.
  - Tagging and releasing branch `3.0` with version `3.0.14`.


Ver 3.2.18 [2023-06-05]:
-------------------------------
  - Only send disposal reminders to necropsy cases


Ver 3.2.16 [2023-06-03]:
-------------------------------
  - Added option to ue text client name


Ver 3.2.14 [2023-05-16]:
-------------------------------


Ver 3.2.12 [2023-05-16]:
-------------------------------


Ver 3.2.10 [2023-05-16]:
-------------------------------


Ver 3.2.8 [2023-05-14]:
-------------------------------
  - Added DKIM signing to institution settings


Ver 3.2.6 [2023-03-02]:
-------------------------------
  - Use replyTo instead of from for emails


Ver 3.2.4 [2022-08-11]:
-------------------------------


Ver 3.2.2 [2022-08-05]:
-------------------------------


Ver 3.2.0 [2022-03-03]:
-------------------------------


Ver 3.0.26 [2022-03-03]:
-------------------------------


Ver 3.0.24 [2021-08-16]:
-------------------------------


Ver 3.0.22 [2021-03-10]:
-------------------------------


Ver 3.0.20 [2021-02-24]:
-------------------------------


Ver 3.0.18 [2020-11-23]:
-------------------------------
  - Tagging and releasing branch `3.0` with version `3.0.14`.


Ver 3.0.14 [2020-10-20]:
-------------------------------


Ver 3.0.12 [2020-02-14]:
-------------------------------


Ver 3.0.10 [2020-01-29]:
-------------------------------


Ver 3.0.8 [2020-01-07]:
-------------------------------


Ver 3.0.6 [2019-05-29]:
-------------------------------


Ver 3.0.4 [2019-05-15]:
-------------------------------


Ver 3.0.2 [2019-05-15]:
-------------------------------


Ver 3.0.0 [2019-05-12]:
-------------------------------


Ver 2.0.26 [2018-10-03]:
-------------------------------


Ver 2.0.24 [2018-09-13]:
-------------------------------
  - Remove Masquerade from App lib


Ver 2.0.22 [2018-09-04]:
-------------------------------


Ver 2.0.20 [2018-08-20]:
-------------------------------


Ver 2.0.18 [2018-08-13]:
-------------------------------


Ver 2.0.16 [2018-07-17]:
-------------------------------
Minor Code Updates


Ver 2.0.14 [2018-07-09]:
-------------------------------
Minor Code Updates


Ver 2.0.12 [2018-05-31]:
-------------------------------
Minor Code Updates


Ver 2.0.11 [2018-05-09]:
-------------------------------


Ver 2.0.10 [2018-05-08]:
-------------------------------


Ver 2.0.9 [2018-05-08]:
-------------------------------
 - Dev.ttek.org


Ver 2.0.8 [2018-01-15]:
-------------------------------
 - Minor Code Updates
 - Near on finished the new status and message system, just needs a debug


Ver 2.0.7 [2017-06-15]:
-------------------------------
 - Minor Code Updates
 - 252s-dev.vet.unimelb.edu.au


Ver 2.0.6 [2017-05-26]:
-------------------------------
 - Minor Code Updates
 - 252s-dev.vet.unimelb.edu.au


Ver 2.0.5 [2017-04-26]:
-------------------------------
 - Minor Code Updates


Ver 2.0.4 [2017-04-27]:
-------------------------------
 - Minor Code Updates
 - 252s-dev.vet.unimelb.edu.au


Ver 2.0.3 [2017-04-02]:
-------------------------------
 - Minor Code Updates
 - 252s-dev.vet.unimelb.edu.au


Ver 2.0.2 [2016-12-30]:
-------------------------------
 - Minor Code Updates
 - 252s-dev.vet.unimelb.edu.au


Ver 2.0.1 [2016-11-11]:
-------------------------------
 - Deprecated debug lib in favor of the framework \Tk\Debug objects and new framework functions folder


Ver 2.0.0 [2016-09-05]:
-------------------------------



config/config.php
<?php
/* Default config options */
$cfg = array();

/*
 * Options mail, smtp, sendmail, qmail
 */
$cfg['mail.driver'] = 'mail';

/*
 * SMTP settings
 */

/**
 * SMTP hosts.
 * Either a single hostname or multiple semicolon-delimited hostnames.
 * You can also specify a different port
 * for each host by using this format: [hostname:port]
 * (e.g. "smtp1.example.com:25;smtp2.example.com").
 * You can also specify encryption type, for example:
 * (e.g. "tls://smtp1.example.com:587;ssl://smtp2.example.com:465").
 * Hosts will be tried in order.
 * @var string
 */
$cfg['mail.smtp.host'] = 'localhost';
// The default SMTP server port.
$cfg['mail.smtp.port'] = 25;
$cfg['mail.smtp.username'] = '';
$cfg['mail.smtp.password'] = '';
// What kind of encryption to use on the SMTP connection. Options: '', 'ssl' or 'tls'
$cfg['mail.smtp.secure'] = '';
// Whether to use SMTP authentication. Uses the Username and Password properties.
$cfg['mail.smtp.enableAuth'] = true;
// Whether to keep SMTP connection open after each message. If this is set to true then to close the connection requires an explicit call to smtpClose().
$cfg['mail.smtp.enableKeepAlive'] = '';


// Checks if the send command was called from this site
$cfg['mail.checkReferer'] = true;
// Add valid domain names as valid referers if needed
$cfg['mail.validReferers'] = '';



/*
 * Other misc options
 * Generally from the site config
 */
// Set this if you want all email to go to this address during debug
//$cfg['system.debug.email'] = '';

// \Tk\Request Used to set the X-Sender-IP, X-Referer headers
//$cfg['request'] = '';
// \Tk\Session Used to set the X-SiteReferer header
//$cfg['session'] = '';

// If true, all email recipients (to, bcc, cc, from) will be set to `system.debug.email` to avoid sending debug emails to live accounts.
// The real recipient addresses will be moved to X-Debug-To,X-Debug-Cc, etc. headers
//$cfg['debug'] = false;

// If set X-Application will be set to this
//$cfg['system.name'] = '';
//$cfg['system.version'] = '';

// Change this to suite your message body encoding
//$cfg['mail.encoding'] = 'UTF-8';


/*
 * DKIM setup
 */
//$cfg['mail.dkim.domain'] = 'example.com';
//$cfg['mail.dkim.private'] = '/path/to/my/private.key';
//$cfg['mail.dkim.private_string'] = ''; // Use the private key string instead of a file

//$cfg['mail.dkim.selector'] = 'phpmailer';      // Default is phpmailer only set if using something else
//$cfg['mail.dkim.selector'] = 'default';      // Default is phpmailer only set if using something else
//$cfg['mail.dkim.passphrase'] = '';



Tk/Mail/Gateway.php
<?php
namespace Tk\Mail;

use Bs\Config;
use \PHPMailer\PHPMailer\PHPMailer;

/**
 * @author Michael Mifsud <http://www.tropotek.com/>
 * @see http://www.tropotek.com/
 * @license Copyright 2016 Michael Mifsud
 */
class Gateway
{

    /**
     * @var array
     */
    protected $params = array();

    /**
     * @var array
     */
    protected $validReferers = array();

    /**
     * @var PHPMailer
     */
    protected $mailer = null;

    /**
     * The status of the last sent message
     * @var bool
     */
    protected $lastSent = null;

    /**
     * @var array
     */
    protected $error = array();

    /**
     * The status of the last sent message
     * @var Message
     */
    protected $lastMessage = null;

    /**
     * @var \Symfony\Component\EventDispatcher\EventDispatcherInterface
     */
    protected $dispatcher = null;

    /**
     * @var string
     */
    protected $host = 'localhost';


    /**
     * Gateway constructor.
     *
     * @param array $params
     * @throws Exception
     */
    public function __construct($params = [])
    {
        // NOTICE: this could be the \Tk\Config object...
        $this->params = $params;
        $this->mailer = new PHPMailer();

        if (isset($this->params['mail.driver'])) {
            // Set the mail driver Default: mail();
            switch ($this->params['mail.driver']) {
                case 'smtp':
                    $this->mailer->isSMTP();
                    $this->mailer->SMTPAuth = $this->params['mail.smtp.enableAuth'];
                    $this->mailer->SMTPKeepAlive = $this->params['mail.smtp.enableKeepAlive'];
                    $this->mailer->SMTPSecure = $this->params['mail.smtp.secure'];
                    $this->mailer->Host = $this->params['mail.smtp.host'];
                    $this->mailer->Port = $this->params['mail.smtp.port'];
                    $this->mailer->Username = $this->params['mail.smtp.username'];
                    $this->mailer->Password = $this->params['mail.smtp.password'];
                    break;
                case 'sendmail':
                    $this->mailer->isSendmail();
                    break;
                case 'qmail':
                    $this->mailer->isQmail();
                    break;
            }
        }

        if (isset($this->params['mail.dkim.domain'])) {
            if (empty($this->params['mail.dkim.domain'])) {
                error_log('Invalid DKIM domain value.');
            }
            if (empty($this->params['mail.dkim.private']) && empty($this->params['mail.dkim.private_string'])) {
                error_log('Invalid DKIM private key value.');
            }

            $this->mailer->DKIM_domain         = $this->params['mail.dkim.domain'] ?? '';
            $this->mailer->DKIM_private        = $this->params['mail.dkim.private'] ?? '';
            $this->mailer->DKIM_private_string = $this->params['mail.dkim.private_string'] ?? '';
            $this->mailer->DKIM_passphrase     = $this->params['mail.dkim.passphrase'] ?? '';
            //$this->mailer->DKIM_selector       = $this->params['mail.dkim.selector'] ?? 'phpmailer';
            $this->mailer->DKIM_selector       = $this->params['mail.dkim.selector'] ?? 'default';
        }

        if (isset($this->params['mail.validReferers'])) {
            $refs = $this->params['mail.validReferers'];
            if (!is_array($refs)) {
                $refs = explode(',',  $refs);
            }
            $this->validReferers = array_merge($this->validReferers, $refs);
        }
        if (isset($_SERVER['HTTP_HOST'])) {
            $this->host = $_SERVER['HTTP_HOST'];
            $this->validReferers[] = array_merge($this->validReferers, array($this->host));
        }
    }

    /**
     *
     * @param Message $message
     * @return bool
     * @throws \Exception
     */
    public function send(Message $message)
    {
        $this->error = array();
        try {
            if (!count($message->getTo())) {
                throw new Exception('No valid recipients found!');
            }
            if (!$message->getFrom()) {
                throw new Exception('No valid sender email found!');
            }
            $this->checkReferer($this->validReferers);

            $event = new MailEvent($this, $message);
            // Dispatch Pre Send Event
            if ($this->dispatcher) {
                $this->dispatcher->dispatch($event, MailEvents::PRE_SEND);
            }

            if ($message->isHtml()) {
                $this->mailer->msgHTML($message->getParsed());
                $this->mailer->AltBody = strip_tags($message->getParsed());
            } else {
                $this->mailer->Body = $message->getParsed();
            }

            $this->mailer->CharSet = 'UTF-8';
            if (isset($this->params['mail.encoding']) && $this->params['mail.encoding']) {
                $this->mailer->CharSet = $this->params['mail.encoding'];
            }

            foreach ($message->getAttachmentList() as $obj) {
                $this->mailer->addStringAttachment($obj->string, $obj->name, $obj->encoding, $obj->type);
            }

            $message->addHeader('X-Application', 'www.tropotek.com');
            $message->addHeader('X-Application-Name', 'www.tropotek.com');
            $message->addHeader('X-Application-Version', '0.0.1');

            if (isset($this->params['system.info.project'])) {
                $message->addHeader('X-Application', $this->params['system.info.project']);
            }
            if (!empty($this->params['site.title'])) {
                $message->addHeader('X-Application-Name', $this->params['site.title']);
            }
            if (isset($this->params['system.info.version'])) {
                $message->addHeader('X-Application-Version', $this->params['system.info.version']);
            }

            /** @var \Tk\Request $request */
            $request = null;
            if (!empty($this->params['request']) && $this->params['request'] instanceof \Tk\Request)
                $request = $this->params['request'];

            if ($request) {
                if ($request->getClientIp())
                    $message->addHeader('X-Sender-IP', $request->getClientIp());
                if ($request->getTkUri()->getHost())
                    $message->addHeader('X-Host', $request->getTkUri()->getHost());
                if ($request->getReferer())
                    $message->addHeader('X-Referer', $request->getReferer()->getRelativePath());
            }

            $this->mailer->Subject = $message->getSubject();

            if (isset($this->params['debug']) && $this->params['debug']) {  // Send dev emails and headers of live emails if testing or debug

                $message->addHeader('X-Debug-To', Message::listToStr($message->getTo()));
                $message->addHeader('X-Debug-From', $message->getFrom());
                if ($message->getReplyTo()) {
                    $message->addHeader('X-Debug-Reply-To', $message->getReplyTo());
                }

                // Set new debug recipient
                $testEmail = 'debug@'.$this->host;
                if (isset($this->params['system.debug.email'])) {
                    $testEmail = $this->params['system.debug.email'];
                    if (is_array($this->params['system.debug.email'])) {
                        foreach ($this->params['system.debug.email'] as $i => $em) {
                            if ($i == 0) {
                                $testEmail = $em;
                                $this->mailer->addAddress($em, 'Debug To');
                            } else {
                                $this->mailer->addCC($em, 'Debug To');
                            }
                        }
                    } else {
                        $this->mailer->addAddress($testEmail, 'Debug To');
                    }
                }

                if ($this->params['system.debug.email.authUser'] && class_exists('\Bs\Config') && Config::getInstance()->getAuthUser()) {
                    $testEmail = Config::getInstance()->getAuthUser()->getEmail();
                    $this->mailer->addAddress($testEmail, Config::getInstance()->getAuthUser()->getName());
                }

                $this->mailer->setFrom($testEmail, 'Debug From');

                if (count($message->getCc())) {
                    $message->addHeader('X-Debug-Cc', Message::listToStr($message->getCc()));
                }
                if (count($message->getBcc())) {
                    $message->addHeader('X-Debug-Bcc', Message::listToStr($message->getBcc()));
                }
            } else {        // Send live emails

                $from = 'noreply@' . $this->host;
                if ($message->getFrom()) $from = $message->getFrom();
                list($e, $n) = Message::splitEmail($from);
                $this->mailer->setFrom($e, $n);

                if ($message->getReplyTo()) {
                    list($e, $n) = Message::splitEmail($message->getReplyTo());
                    $this->mailer->addReplyTo($e, $n);
                }

                foreach ($message->getTo() as $email) {
                    list($e, $n) = Message::splitEmail($email);
                    $this->mailer->addAddress($e, $n);
                }
                foreach ($message->getCc() as $email) {
                    list($e, $n) = Message::splitEmail($email);
                    $this->mailer->addCC($e, $n);
                }
                foreach ($message->getBcc() as $email) {
                    list($e, $n) = Message::splitEmail($email);
                    $this->mailer->addBCC($e, $n);
                }
            }

            // Set dkim identity
            if (isset($this->params['mail.dkim'])) {
                $this->mailer->DKIM_identity = $this->mailer->From;
            }

            foreach ($message->getHeadersList() as $h => $v) {
                $this->mailer->addCustomHeader($h, $v);
            }

            // Send Email
            $this->lastMessage = $message;

            // Note: can interfear with output buffer contents in AJAX calls, so only enable when testing
//            if (\Tk\Config::getInstance()->isDebug())
//                $this->mailer->SMTPDebug = 2;

            $this->lastSent = $this->mailer->send();
            if (!$this->lastSent)
                throw new \Tk\Mail\Exception($this->mailer->ErrorInfo);

            // Dispatch Post Send Event
            if ($this->dispatcher) {
                $this->dispatcher->dispatch($event, MailEvents::POST_SEND);
            }

        } catch (\Exception $e) {
            // TODO: Discuss if this is the best way or should we catch exceptions externally, that may be a better option...???????s
            $this->lastSent = false;
            $this->error[] = $e->getMessage();
            throw $e;
        }

        $this->mailer->clearAllRecipients();
        $this->mailer->clearAttachments();
        $this->mailer->clearCustomHeaders();
        $this->mailer->clearReplyTos();
        return $this->lastSent;
    }

    /**
     * @return array
     */
    public function getErrors()
    {
        return $this->error;
    }

    /**
     * Get the last sent message status
     *
     * @return bool
     */
    public function getLastSent()
    {
        return $this->lastSent;
    }

    /**
     * Return the last message that was sent
     *
     * @return Message
     */
    public function getLastMessage()
    {
        return $this->lastMessage;
    }

    /**
     * @return \Symfony\Component\EventDispatcher\EventDispatcherInterface
     */
    public function getDispatcher()
    {
        return $this->dispatcher;
    }

    /**
     * @param \Symfony\Component\EventDispatcher\EventDispatcherInterface $dispatcher
     */
    public function setDispatcher($dispatcher)
    {
        $this->dispatcher = $dispatcher;
    }

    /**
     * @return PHPMailer
     */
    public function getMailer()
    {
        return $this->mailer;
    }

    /**
     * See if a string contains any suspicious/injection coding.
     *
     * @param string $str
     * @throws \Tk\Mail\Exception
     */
//    private function validateString($str)
//    {
//        if (!$str) { return; }
//        $badStrings = array("content-type:", "mime-version:", "multipart\/mixed", "content-transfer-encoding:", "bcc:", "cc:", "to:");
//        foreach ($badStrings as $badString) {
//            if (preg_match('/'.$badString.'/i', strtolower($str))) {
//                throw new Exception("'$badString' found. Suspected injection attempt - mail not being sent.");
//            }
//        }
//        if (preg_match("/(%0A|%0D|\\n+|\\r+)/i", $str) != 0) {
//            throw new Exception("newline found in '$str'. Suspected injection attempt - mail not being sent.");
//        }
//    }

    /**
     * check_referer() breaks up the environmental variable
     * HTTP_REFERER by "/" and then checks to see if the second
     * member of the array (from the explode) matches any of the
     * domains listed in the $referers array (declared at top)
     *
     * @param array $referers
     * @throws \Tk\Mail\Exception
     */
    private function checkReferer($referers)
    {
        // do not check referrer for CLI apps
        if (substr(php_sapi_name(), 0, 3) == 'cli') {
            return;
        }
        if (!isset($this->params['mail.checkReferer']) || !$this->params['mail.checkReferer']) {
            return;
        }

        if (count($referers) > 0) {
            if (isset($_SERVER['HTTP_REFERER'])) {
                $temp = explode('/', $_SERVER['HTTP_REFERER']);
                $found = false;
                foreach ($referers as $k => $stored_referer) {
                    if (preg_match('/^' . $stored_referer . '$/i', $temp[2])) {
                        $found = true;
                        break;
                    }
                }
//                while (list(, $stored_referer) = each($referers)) {
//                    if (preg_match('/^' . $stored_referer . '$/i', $temp[2]))
//                        $found = true;
//                }
                if (!$found) {
                    throw new Exception("You are coming from an unauthorized domain. Illegal Referer.");
                }
            } else {
                throw new Exception("Sorry, but I cannot figure out who sent you here. Your browser is not sending an HTTP_REFERER. This could be caused by a firewall or browser that removes the HTTP_REFERER from each HTTP request you submit.");
            }
        } else {
            throw new Exception("There is no referer defined. All submissions will be denied.");        }
    }

}

Tk/Mail/CurlyMessage.php
<?php
namespace Tk\Mail;

use Tk\Callback;

/**
 * This message accepts text templates and replaces {param} with the
 * corresponding value.
 *
 *
 *
 */
class CurlyMessage extends Message
{
    
    use \Tk\CollectionTrait;

    /**
     * @var Callback
     */
    protected $onParse = null;

    /**
     * @var \Tk\CurlyTemplate
     */
    protected $template = null;


    /**
     * MessageTemplate constructor.
     * 
     * @param string $body
     * @param string $subject
     * @param string $to
     * @param string $from
     */
    public function __construct($body = '{content}', $subject = '', $to = '', $from = '')
    {
        parent::__construct($body, $subject, $to, $from);
        $this->set('content', '');
    }

    /**
     *  Do a deep clone to avoid template issues
     */
    function __clone()
    {
        if ($this->template)
            $this->template = clone $this->template;
        if ($this->collection)
            $this->collection = clone $this->collection;
    }

    /**
     * Set the content. this should be the contents of the email
     * not to be confused with the message template.
     * It can contain curly template vars also.
     *
     * @param string $tpl
     */
    public function setContent($tpl)
    {
        $this->set('content', $tpl);
    }

    /**
     * The message text body
     *
     * @param string $body
     * @return $this
     */
    public function setBody($body)
    {
        $this->body = $body;
        $this->template = null;
        if ($body)
            $this->template = \Tk\CurlyTemplate::create($body);
        return $this;
    }

    /**
     * @return Callback
     */
    public function getOnParse()
    {
        return $this->onParse;
    }

    /**
     * Set a callback function to fire when the getParsed() method is called
     * EG: function ($curlyMessage) { }
     *
     * @param callable $callable
     * @param int $priority
     * @return $this
     */
    public function addOnParse($callable, $priority=Callback::DEFAULT_PRIORITY)
    {
        $this->getOnParse()->append($callable, $priority);
        return $this;
    }

    /**
     * Set a callback function to fire when the getParsed() method is called
     *
     * @param callable $onParse
     * @return $this
     * @deprecated Use addOnParse($callable, $priority);
     */
    public function setOnParse($onParse)
    {
        $this->addOnParse($onParse);
        return $this;
    }

    /**
     * Gets the tCurlyTemplate object
     *
     * This will return null until the setBody($body) function is called with data
     *
     * @return \Tk\CurlyTemplate
     */
    public function getTemplate()
    {
        return $this->template;
    }

    /**
     * Returns the a parsed message body ready for sending.
     *
     * @return string
     * @throws \Tk\Exception
     */
    public function getParsed()
    {
        if (!$this->template) return '';

        $this->set('subject', $this->getSubject());
        $this->set('fromEmail', $this->getFrom());
        $this->set('toEmail', self::listToStr($this->getTo()));
        $this->set('toEmailList', self::listToStr($this->getTo()));
        $this->set('ccEmailList', self::listToStr($this->getCc()));
        $this->set('bccEmailList', self::listToStr($this->getBcc()));
        $this->set('date', \Tk\Date::create()->format(\Tk\Date::FORMAT_LONG_DATETIME));

        if (is_callable($this->onParse)) {
            call_user_func_array($this->onParse, array($this));
        }

        $str = $this->template->parse($this->getCollection()->all());

        return $str;
    }


    /**
     * @return array
     */
    public static function getParamList()
    {
        return array(
            'subject',
            'fromEmail',
            'toEmail',
            'toEmailList',
            'ccEmailList',
            'bccEmailList'
        );
    }

}

config/20-config.php
<?php
/**
 * Setup phpmailer configuration parameters
 *
 * @author Tropotek <http://www.tropotek.com/>
 */
use Tk\Config;

return function (Config $config)
{

    /*
     * Options (php)mail, smtp, sendmail, qmail
     */
    $config->set('mail.driver', 'mail');

    /*
     * SMTP settings
     */

    /**
     * SMTP hosts.
     * Either a single hostname or multiple semicolon-delimited hostnames.
     * You can also specify a different port
     * for each host by using this format: [hostname:port]
     * (e.g. "smtp1.example.com:25;smtp2.example.com").
     * You can also specify encryption type, for example:
     * (e.g. "tls://smtp1.example.com:587;ssl://smtp2.example.com:465").
     * Hosts will be tried in order.
     */
    $config->set('mail.smtp.host', 'localhost');
    $config->set('mail.smtp.port', 25);
    $config->set('mail.smtp.username', '');
    $config->set('mail.smtp.password', '');
    // What kind of encryption to use on the SMTP connection. Options: '', 'ssl' or 'tls'
    $config->set('mail.smtp.secure', '');
    // Whether to use SMTP authentication. Uses the Username and Password properties.
    $config->set('mail.smtp.enableAuth', true);
    // Whether to keep SMTP connection open after each message.
    //   If this is set to true then to close the connection requires an explicit call to smtpClose().
    $config->set('mail.smtp.username', false);

    // Checks if the send command was called from this site
    $config->set('mail.checkReferer', true);
    // Add valid domain names as valid referrers if needed
    $config->set('mail.validReferrers', '');

    /*
     * Other misc options
     */

    // Set this if you want all email to go to this address in debug mode
    //$config->set('mail.debug.email', 'user@example.com');

    // If set X-Application will be set to this
    //$config->set('mail.name', 'tk-mail');
    //$config->set('mail.version', ''1.0.0);

    // Change this to suite your message body encoding
    $config->set('mail.encoding', 'UTF-8');


    // DKIM SETUP

    //$config->set('mail.dkim.domain', 'example.com');
    //$config->set('mail.dkim.private', '/path/to/my/private.key');
    //$config->set('mail.dkim.private_string', '');             // Use the private key string instead of a file

};



